File: C++/string_compression.cpp
/*
 * Cracking the coding interview edition 6
 * Problem 1-6 Implement a method to perform basic string compression.
 * Example string aabcccccaaa should be compressed to a2b1c5a3,
 * however if compressed string is bigger than original string, return original string
 */

#include <iostream>
#include <numeric>
#include <string>
#include <sstream>
#include <tuple>

// This function takes a string as input and compresses it using a simple algorithm
std::string compress(std::string str)
{
    // If the length of the string is less than 2, there is nothing to compress
    if (str.length() < 2) return str;

    // Create an output stringstream to store the compressed string
    std::stringstream out;

    // Initialize a counter to keep track of the number of consecutive characters
    int count = 1;

    // Iterate over the characters of the string starting from the second character
    for (auto it = str.begin() + 1; it <= str.end(); ++it) {
        // If the current character is different from the previous character, write the previous character
        // and the count to the output stringstream, reset the counter, and continue iterating
        if (*it != *(it-1)) {
            out << *(it-1) << count;
            count = 1;
        }
        // If the current character is the same as the previous character, increment the counter
        else {
            ++count;
        }
    }

    // Check if the length of the compressed string is less than the length of the input string
    // If it is, return the compressed string. Otherwise, return the input string.
    return out.str().length() < str.length() ? out.str() : str;
}


int main() {
    std::string str;
    std::cout << "Enter a string: ";
    std::cin >> str;

    auto compressed = compress(str);
    std::cout << str << (str != compressed ? " can be compressed to " + compressed : " can not be compressed") << std::endl;

    return 0;
}

File: Python/string_compression.py
import time
import unittest
from itertools import groupby

def compress1(s):
    # Check if the length of the string is less than 2
    if len(s) < 2:
        return s

    # Initialize variables to store compressed string and count of consecutive characters
    compressed = ""
    count = 1
    
    # Iterate through the string
    for i in range(1, len(s)+1):
        # If the current character is different from the previous character, append the previous character and its count to the compressed string
        if ((i==len(s)) or (s[i] != s[i-1])):
            compressed += s[i-1] + str(count)
            count = 1
        # If the current character is the same as the previous character, increment the count
        else:
            count += 1
    
    # If the length of the compressed string
    return compressed if len(compressed) < len(s) else s





def compress2(s):
    # Group consecutive identical characters using itertools.groupby
    groups = groupby(s)
    
    # Create a list of tuples containing the label and count of each group
    result = [(label, len(list(group))) for label, group in groups]
    
    # Concatenate the label and count of each group into a compressed string
    compressed = "".join([f"{label}{count}" for label, count in result])
    
    # Return the compressed string if it's shorter than the original string, otherwise return the original string
    return compressed if len(compressed) < len(s) else s




class Test(unittest.TestCase):
    test_cases = [
        ("aabcccccaaa", "a2b1c5a3"),
        ("abcdef", "abcdef"),
        ("aabb", "aabb"),
        ("aaa", "a3"),
        ("a", "a"),
        ("", ""),
    ]
    testable_functions = [
        compress1,compress2
    ]

    def test(self):
        for test_function in self.testable_functions:
            for args, expected in self.test_cases:
                assert test_function(args) == expected, f"Failed {test_function.__name__} for: {[*args]}"


if __name__ == "__main__":
    unittest.main()